extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java

```diff
@@ -137,205 +103,44 @@ public List<WritableFrameChannel> outputChannels()
-     * Frame 2 -> rac2
-     * 3, 1 --> key changed
-     * 3, 2
-     * 3, 3
-     * 3, 4
-     *
-     * Frame 3 -> rac3
-     *
-     * 3, 5
-     * 3, 6
-     * 4, 1 --> key changed
-     * 4, 2
-     *
-     * In case of empty OVER clause, all these racs need to be added to a single rows and columns
-     * to be processed. The way we can do this is to use a ConcatRowsAndColumns
-     * ConcatRC [rac1, rac2, rac3]
-     * Run all ops on this
-     *
-     *
-     * The flow would look like:
-     * 1. Validate if the operator doesn't have any OVER() clause with PARTITION BY for this stage.
-     * 2. If 1 is true make a giant rows and columns (R&C) using concat as shown above
-     *    Let all operators run amok on that R&C
-     * 3. If 1 is false
-     *    Read a frame
-     *    keep the older row in a class variable
-     *    check row by row and compare current with older row to check if partition boundary is reached
-     *    when frame partition by changes
-     *    create R&C for those particular set of columns, they would have the same partition key
-     *    output will be a single R&C
-     *    write to output channel
-     *
-     *
+     * We add all the RACs to a single ConcatRowsAndColumns and pass it to the list of operators to process.
      *  Future thoughts: {@link https://github.com/apache/druid/issues/16126}
-     *
-     *  1. We are writing 1 partition to each frame in this way. In case of high cardinality data
-     *      we will be making a large number of small frames. We can have a check to keep size of frame to a value
-     *      say 20k rows and keep on adding to the same pending frame and not create a new frame
-     *
-     *  2. Current approach with R&C and operators materialize a single R&C for processing. In case of data
-     *     with low cardinality a single R&C might be too big to consume. Same for the case of empty OVER() clause
-     *     Most of the window operations like SUM(), RANK(), RANGE() etc. can be made with 2 passes of the data.
-     *     We might think to reimplement them in the MSQ way so that we do not have to materialize so much data
      */
 
-    if (partitionColumnNames.isEmpty()) {
-      // If we do not have any OVER() clause with PARTITION BY for this stage.
-      // Bring all data to a single executor for processing.
-      // Convert each frame to RAC.
-      // Concatenate all the racs to make a giant RAC.
-      // Let all operators run on the giant RAC until channel is finished.
-      if (inputChannel.canRead()) {
-        final Frame frame = inputChannel.read();
-        convertRowFrameToRowsAndColumns(frame);
-      } else if (inputChannel.isFinished()) {
-        runAllOpsOnMultipleRac(frameRowsAndCols);
-        return ReturnOrAwait.returnObject(Unit.instance());
-      } else {
-        return ReturnOrAwait.awaitAll(inputChannels().size());
-      }
-      return ReturnOrAwait.runAgain();
+    // Bring all data to a single executor for processing.
```

Comment:
Probably a few things to check : 
* Should we adjust the num rows per partition to some X number if window query is encountered. 
* What are the guardrails if we go OOM. Can we fail gracefully instead of getting an annoying OOM error.

Reply:
@cryptoe Regarding the 2nd point: We have a query context parameter `maxRowsMaterializedInWindow`, and the following guardrail:

```java
if (rac.numRows() > maxRowsMaterialized) {
       throw new MSQException(new TooManyRowsInAWindowFault(rac.numRows(), maxRowsMaterialized));
     }
```

The default value of the above param is set to 100k in `Limits.java` file:
```java
public static final int MAX_ROWS_MATERIALIZED_IN_WINDOW = 100_000;
```

Reply:
> Should we adjust the num rows per partition to some X number if window query is encountered.

@cryptoe Do you mean by increasing the number of partitions in shuffleSpec of the previous stage? If not, can you please elaborate?

Reply:
I believe the current approach will enforce that at most a window with `MAX_ROWS_MATERIALIZED_IN_WINDOW` can be processed (on a single worker). I believe later we will enhance this to be able to stream these things - which will naturally relax these checks a bit.

I believe we are doing everything for the execution engine by specifying the columns it may spread the data - if the partition number or something should be increased there; I believe that shouldn't be visible around here as other pieces may benefit from such a logic.

I think if we want to improve on that - we could possibly do that in a separate followup.

Reply:
@cryptoe @kgyrtkirk I've made the changes as per https://github.com/apache/druid/pull/16823#issuecomment-2292772397, please have a look. Thanks!

Closing this comment as it's not relevant anymore to the latest changes of the PR.